### ISSUE TITLE
decrement for counter_cache breaks in nested attribute

### ISSUE THREAD

**Issue by otux** [rails/rails#27169]
**Problem:**
The decrement of the counter cache is not triggered correctly, when connected via nested attribute, so that in an after_save method wrong values are the result.

********************************************************
My scenario consists (in part) of the following (as I assume) relevant classes:

```ruby
class Event < ActiveRecord::Base
  
  after_save :update_posted_categories
  has_many :event_dates, inverse_of: :event
  accepts_nested_attributes_for :event_dates, allow_destroy: true

  private

  def update_posted_categories
    p "#{event_dates_count_was} -> #{event_dates_count}"
  end
end

class EventDate < ActiveRecord::Base
  belongs_to :event, inverse_of: :event_dates, counter_cache: true
end
```

The SQL that is executed when removing an event_date reads:

SQL (0.6ms)  UPDATE "events" SET "event_dates_count" = COALESCE("event_dates_count", 0) - 1 WHERE "events"."id" = $1  [["id", 2629]]

but the "p" in the update_posted_categories call returns:
"21 -> 21"

whereas it actually should read:
"21 -> 20"

********************************************************
**'Fix':**

I made my way into the file
activerecord-4.2.6/lib/active_record/associations/belongs_to_association.rb

where I changed:
```ruby
        def decrement_counter(counter_cache_name)
          if foreign_key_present?
            klass.decrement_counter(counter_cache_name, target_id)
          end
        end
```
into:
```ruby
        def decrement_counter(counter_cache_name)
          if foreign_key_present?
            klass.decrement_counter(counter_cache_name, target_id)
            if target && !stale_target? && counter_cache_available_in_memory?(counter_cache_name)
              target.decrement(counter_cache_name)
            end
          end
        end
```
(The if-block is copied over from the following **in**crement_counter definition)
This fixes the problem scenario for me. Shouldn't that be there in the first place?

**Comment by utilum** [comment]
@otux please clarify:
1) Rails version (evidently it isn't the [current version](http://api.rubyonrails.org/classes/ActiveRecord/CounterCache/ClassMethods.html#method-i-increment_counter)).
2) Why do you expect the counter to decrease when saving an event?
3) How come you fix the problem by modifying a method not called in your code?
The last two would be best served by an executable bug report for your version of Rails.

**Comment by mck9** [comment]
@utilum 
1. Version 4.2-head
2. The EventDate is removed through nested attributes
3. The method is called through the counter_cache option in "belongs_to :event, inverse_of: :event_dates, counter_cache: true"

The counter_cache "event_dates_count" works as expected when EventDates are created. The increment method is in rails/activerecord/lib/active_record/associations/belongs_to_association.rb defined as:
```ruby
        def increment_counter(counter_cache_name)
          if foreign_key_present?
            klass.increment_counter(counter_cache_name, target_id)
            if target && !stale_target? && counter_cache_available_in_memory?(counter_cache_name)
              target.increment(counter_cache_name)
            end
          end
        end
```

**Comment by utilum** [comment]
Thank you for clarifying.
Does [this executable script](https://gist.github.com/utilum/5296f9e34ee8bcfaf8a5841164326968) reproduce the case?

**Comment by utilum** [comment]
@otux this has nothing to do with the association being nested or not, but does show different behavior between 4.7.2 and >= 5. 

In version >= 5 it depends on how you "remove" the association, counter is not decremented after `delete`, *is* decremented after `destroy`. Probably (?) expected behavior.

In 4.7.2 it is not updated on either, and I doubt that's expected. I've updated the [executable test script](https://gist.github.com/utilum/5296f9e34ee8bcfaf8a5841164326968) accordingly.

Normalizing `decrement_counter` and `increment_counter` as you propose does fix this, and indeed it's been [generalized](https://github.com/rails/rails/blob/v5.0.0/activerecord/lib/active_record/associations/belongs_to_association.rb#L44) in v5.

Do you want to make a PR to port this? (Or I might).

**Comment by utilum** [comment]
@maclover7 is backporting this change to 4-2-stable desirable?

**Comment by otux** [comment]
@utilum 
as I understand it, there is no difference between 4.7.2 and >=5 regarding callbacks;
according to http://api.rubyonrails.org/classes/ActiveRecord/Relation.html#method-i-delete delete is not expected to trigger a callback;
therefore I made two changes to your script, marked with 'CHANGE' (lines 111 and 138), because I would not expect 0 (as you wrote) but 1;
I added the patch for the decrement_counter for the active_record to the test script as well (see 'PATCH' in script file);
with the patch all (changed) tests run through as I would expect them to;
**Changed test script see:** https://gist.github.com/otux/43b963cc48312a6099a8d7e8ed469522

**Comment by utilum** [comment]
Good show, @otux . Want to make a PR of it?

**Comment by utilum** [comment]
Mind you, the way they nicely do it in 5 is by extracting both `decrement_counter` and `increment_counter` methods to a private `update_counter`. So you'd probly like to follow that example, and add tests.

**Comment by maclover7** [comment]
@utilum good question, not the best person to answer this, so cc @sgrif

**Comment by sgrif** [comment]
This was a case where backporting had some subtle changes in behavior that I wasn't comfortable with making. Officially we don't actually support 4.2 for bug fixes any more (though we try to when it's easy for us to do). Since this is fixed on 5.0, there is no bug.
